src/java.base/share/classes/jdk/internal/classfile/impl/verifier/ParserVerifier.java

```diff
@@ -0,0 +1,371 @@
+import static java.lang.constant.ConstantDescs.INIT_NAME;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.lang.classfile.Attribute;
+import java.lang.classfile.AttributedElement;
+import java.lang.classfile.Attributes;
+import java.lang.classfile.ClassFile;
+import java.lang.classfile.ClassModel;
+import java.lang.classfile.ClassFileElement;
+import java.lang.classfile.CodeModel;
+import java.lang.classfile.CompoundElement;
+import java.lang.classfile.FieldModel;
+import java.lang.classfile.MethodModel;
+import java.lang.classfile.TypeKind;
+import java.lang.classfile.attribute.*;
+import java.lang.classfile.constantpool.*;
+import java.lang.constant.ConstantDescs;
+import java.lang.reflect.AccessFlag;
+import jdk.internal.classfile.impl.BoundAttribute;
+import jdk.internal.classfile.impl.Util;
+
+/**
+ * ParserVerifier performs selected checks of the class file format according to
+ * {@jvms 4.8 Format Checking}
+ *
+ * @see <a href="https://raw.githubusercontent.com/openjdk/jdk/master/src/hotspot/share/classfile/classFileParser.cpp">hotspot/share/classfile/classFileParser.cpp</a>
+ */
+public record ParserVerifier(ClassModel classModel) {
+
+    List<VerifyError> verify() {
+        var errors = new ArrayList<VerifyError>();
+        verifyConstantPool(errors);
+        verifyInterfaces(errors);
+        verifyFields(errors);
+        verifyMethods(errors);
+        verifyAttributes(classModel, errors);
+        return errors;
+    }
+
+    private void verifyConstantPool(List<VerifyError> errors) {
+        for (var cpe : classModel.constantPool()) {
+            Consumer<Runnable> check = c -> {
+                try {
+                    c.run();
+                } catch (VerifyError|Exception e) {
+                    errors.add(new VerifyError("%s at constant pool index %d in %s".formatted(e.getMessage(), cpe.index(), toString(classModel))));
+                }
+            };
+            check.accept(switch (cpe) {
+                case DoubleEntry de -> de::doubleValue;
+                case FloatEntry fe -> fe::floatValue;
+                case IntegerEntry ie -> ie::intValue;
+                case LongEntry le -> le::longValue;
+                case Utf8Entry ue -> ue::stringValue;
+                case ConstantDynamicEntry cde -> cde::asSymbol;
+                case InvokeDynamicEntry ide -> ide::asSymbol;
+                case ClassEntry ce -> ce::asSymbol;
+                case StringEntry se -> se::stringValue;
+                case MethodHandleEntry mhe -> mhe::asSymbol;
+                case MethodTypeEntry mte -> mte::asSymbol;
+                case FieldRefEntry fre -> {
+                    check.accept(fre.owner()::asSymbol);
+                    check.accept(fre::typeSymbol);
+                    yield () -> verifyFieldName(fre.name().stringValue());
```

Comment:
Nitpick, we should avoid capturing the check instance and just do something like:
```java
                case FieldRefEntry fre -> () -> {
                    fre.owner().asSymbol();
                    fre.typeSymbol();
                    verifyFieldName(fre.name().stringValue());
                };
```

Reply:
The above code logs all errors individually.